Docs: document options and trade-offs to query label names and values (#5582)
grafana/mimir: docs/sources/mimir/query/query-metric-labels.md

---
description: Learn how to query metric labels from Grafana Mimir.
title: Query metric labels
menuTitle: Query metric labels
weight: 2
keywords:
  - labels
---

# Query metric labels

Grafana Mimir supports multiple [HTTP API]({{< relref "../references/http-api/index.md" >}}) endpoints to query metric label names and values.

> In this document we explain the main differences and trade-off between these API endpoints.

Reviewer: ```suggestion
There are differences and trade-off between these API endpoints.
```

Reviewer: Rationale: Eliminate talking about the meta layer of documentation; convey the concepts directly.

## Querying label names

> The following API endpoints can be used to find the list of label names of all (or a subset of) series stored in Mimir:

Reviewer: ```suggestion
The following API endpoints find the list of label names of all (or a subset of) series stored in Mimir:
```

- **[Get label names]({{< relref "../references/http-api/index.md#get-label-names" >}})**<br />
  `GET,POST <prometheus-http-prefix>/api/v1/labels`
- **[Label names cardinality]({{< relref "../references/http-api/index.md#label-names-cardinality" >}})**<br />
  `GET,POST <prometheus-http-prefix>/api/v1/cardinality/label_names`

> > **Note**: `<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.
> > The default prefix is `/prometheus`.

Reviewer: ```suggestion
{{% admonition type="note" %}}
`<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.
The default prefix is `/prometheus`.{{% /admonition %}}
```

Reviewer: This change is recent. For guidance, refer to https://grafana.com/docs/writers-toolkit/write/shortcodes/#admonition-shortcode.

### Recommendations and performance trade-off

> - Prefer "get label names" over "label names cardinality" API, unless you need to know the number of unique label values for each label name.

Reviewer: ```suggestion
- Prefer _get label names_ over the _label names cardinality_ API, unless you need to know the number of unique label values for each label name.
```

> - The "get label names" API `start` and `end` parameters are optional. When omitted, Mimir will query label names for the whole retention period or up to the configured `-store.max-labels-query-length`. We recommend to always specify `start` and `end` parameters, and preferably set to the shortest period which is feasible for your use case.

Reviewer: ```suggestion
- The _get label names_ API `start` and `end` parameters are optional. When omitted, Mimir will query label names for the whole retention period or up to the configured `-store.max-labels-query-length`. We recommend to always specify `start` and `end` parameters, and preferably set them to the shortest period that is feasible for your use case.
```

> - For both API endpoints, executing a request specifying the series selector is computationally more expensive than the same request without the series selector, but the results set can be significantly smaller when the series selector effectively reduce the series set.

Reviewer: ```suggestion
- For both API endpoints, executing a request specifying the series selector is computationally more expensive than the same request without the series selector. However, the results set might be significantly smaller when the series selector effectively reduces the series set.
```
Can we remove "effectively" here? Do you mean "successfully"?

Author: We can remove it.

### Features comparison

> The different API endpoints have different features.
> The following list summarises the main differences.

Reviewer: ```suggestion
The different API endpoints have different features,
and the main differences follow.
```

#### Can specify the time range in the request?

- **Get label names**<br />

>   Yes. However, the actual time range granularity is restricted to TSDB block range periods (2h for the most recent metrics and up to 24h for historical metrics when running Mimir with the default configuration). In practice, the input `start` and `end` parameters are rounded to the TSDB block boundaries: the `start` parameter is rounded to the start of the nearest block containing samples whose timestamp is lower or equal than the input timestamp, while the `end` parameter is rounded to the end of the nearest block containing samples with timestamp greater or equal than the input timestamp.

Reviewer: ```suggestion
  Yes. However, the actual time range granularity is restricted to TSDB block range periods, which are `2h` for the most recent metrics, and up to `24h` for historical metrics when running Mimir with the default configuration.
  In practice, the input `start` and `end` parameters are rounded to the TSDB block boundaries:
  * The `start` parameter is rounded to the start of the nearest block containing samples with timestamps that are less than or equal to the input timestamp.
  * The `end` parameter is rounded to the end of the nearest block containing samples with timestamps that are greater than or equal to the input timestamp.
```
This might need some more verbiage tweaks. I am not sure if I have changed the technical accuracy in my suggestions.

Author: It's still technically accurate 👍

- **Label names cardinality**<br />
  No. The API queries only the in-memory series in the ingesters that hold series for the tenant.

#### Can specify the series selector in the request?

- **Get label names API**<br />

>   Yes. The optional `match[]` parameter allows to select the series from which to read the label names. The `match[]` parameter can be specified multiple times. When omitted, label names are extracted from all series.

Reviewer: ```suggestion
  Yes. The optional `match[]` parameter makes it possible to select the series from which to read the label names. The `match[]` parameter can be specified multiple times. When omitted, label names are extracted from all series.
```

- **Label names cardinality API**<br />

>   Yes. The optional `selector` parameter allows to select the series from which to read the label names. The `selector` parameter can be specified only once. When omitted, label names are extracted from all series.

Reviewer: ```suggestion
  Yes. The optional `selector` parameter makes it possible to select the series from which to read the label names. The `selector` parameter can be specified only once. When omitted, label names are extracted from all series.
```

## Querying label values

The following API endpoints can be used to find the list of label values for a given label name:

- **[Get label values]({{< relref "../references/http-api/index.md#get-label-values" >}})**<br />
  `GET <prometheus-http-prefix>/api/v1/label/{name}/values`
- **[Label values cardinality]({{< relref "../references/http-api/index.md#label-values-cardinality" >}})**<br />
  `GET,POST <prometheus-http-prefix>/api/v1/cardinality/label_values`

> > **Note**: `<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.

Reviewer: ```suggestion
{{% admonition type="note" %}}
`<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.
{{% /admonition %}}
```

> 
> > **Note**: `<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.
> > The default prefix is `/prometheus`.

Reviewer: ```suggestion

{{% admonition type="note" %}}
`<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.
The default prefix is `/prometheus`.
{{% /admonition %}}
```

- **Label names cardinality API**<br />
  Yes. The optional `selector` parameter allows to select the series from which to read the label names. The `selector` parameter can be specified only once. When omitted, label names are extracted from all series.

## Querying label values

The following API endpoints find the list of label values for a given label name:

- **[Get label values]({{< relref "../references/http-api/index.md#get-label-values" >}})**<br />
  `GET <prometheus-http-prefix>/api/v1/label/{name}/values`
- **[Label values cardinality]({{< relref "../references/http-api/index.md#label-values-cardinality" >}})**<br />
  `GET,POST <prometheus-http-prefix>/api/v1/cardinality/label_values`

{{% admonition type="note" %}}
`<prometheus-http-prefix>` is the Prometheus HTTP prefix as documented in the [HTTP API]({{< relref "../references/http-api/index.md#path-prefixes" >}}) reference.
The default prefix is `/prometheus`.
{{% /admonition %}}

### Recommendations and performance trade-off

- Prefer _get label values_ over _label values cardinality_ API, unless you need to know the number of series for each label name and value.

> - The _get label values_ API allows to specify a single label name, while the _label values cardinality_ API allows to specify multiple label names. If you need to get the label values for multiple label names, we recommend issuing multiple _get label values_ requests instead of a single _label values cardinality_ API call with multiple label names.

Reviewer: ```suggestion
- The _get label values_ API allows to specify a single label name, and the _label values cardinality_ API allows to specify multiple label names. If you need to get the label values for multiple label names, we recommend issuing multiple _get label values_ requests instead of a single _label values cardinality_ API call with multiple label names.
```

Reviewer: "while¨ means "and" in this context.
